addons/hr_work_entry/models/hr_work_entry.py

```diff
@@ -77,19 +78,36 @@ def _compute_conflict(self):
 
     @api.depends('date_stop', 'date_start')
     def _compute_duration(self):
+        durations = self._get_duration_batch()
         for work_entry in self:
-            work_entry.duration = work_entry._get_duration(work_entry.date_start, work_entry.date_stop)
+            work_entry.duration = durations[work_entry.id]
 
     @api.depends('date_start', 'duration')
     def _compute_date_stop(self):
         for work_entry in self.filtered(lambda w: w.date_start and w.duration):
             work_entry.date_stop = work_entry.date_start + relativedelta(hours=work_entry.duration)
 
+    def _get_duration_batch(self):
+        result = {}
+        cached_periods = defaultdict(lambda: 0.0)
+        for work_entry in self:
+            date_start = work_entry.date_start
+            date_stop = work_entry.date_stop
+            if not date_start or not date_stop:
+                result[work_entry.id] = 0.0
+                continue
+            if (date_start, date_stop) in cached_periods:
+                result[work_entry.id] = cached_periods[(date_start, date_stop)]
+            else:
+                dt = date_stop - date_start
+                duration = dt.days * 24 + dt.seconds / 3600  # Number of hours
+                cached_periods[(date_start, date_stop)] = duration
```

Comment:
Is this cache on dt content really useful for the perf? :eyes:

Reply:
Well, it seems to be half the execution time. In our use case, 99.99 % of the intervals have the same duration (same hours)

```py
import time
from datetime import datetime
from collections import defaultdict

# Number of periods : 1000 employees x 22 days per months x 2 periods (morning/afternoon)
start = time.time()
for i in range(1000):
    for j in range(1, 23):
        for k in [(datetime(2021, 1, j, 8), datetime(2021, 1, j, 12)), (datetime(2021, 1, j, 13), datetime(2021, 1, j, 18))]:
            dt = k[1] - k[0]
            duration = dt.days * 24 + dt.seconds / 3600

time.time() - start

start = time.time()
cached_periods = defaultdict(lambda: 0.0)
for i in range(1000):
    for j in range(1, 23):
        for k in [(datetime(2021, 1, j, 8), datetime(2021, 1, j, 12)), (datetime(2021, 1, j, 13), datetime(2021, 1, j, 18))]:
            if (k[0], k[1]) in cached_periods:
                duration = cached_periods[(k[0], k[1])]
            else:
                dt = k[1] - k[0]
                duration = dt.days * 24 + dt.seconds / 3600  # Number of hours
                cached_periods[(k[0], k[1])] = duration

time.time() - start
```
Returns 0.07 and 0.03 seconds for example.